File: src/main/scala/io/snyk/plugin/ui/settings/SettingsDialog.scala
package io.snyk.plugin.ui.settings

import io.snyk.plugin.Utils.isUrlValid
import java.util.Objects.nonNull
import java.awt.{Dimension, Insets}

import com.intellij.openapi.Disposable
import com.intellij.openapi.ui.{ComponentValidator, ValidationInfo}
import com.intellij.ui.components.fields.ExpandableTextField
import com.intellij.ui.{DocumentAdapter, IdeBorderFactory}
import com.intellij.uiDesigner.core.{Spacer, GridConstraints => UIGridConstraints, GridLayoutManager => UIGridLayoutManager}
import javax.swing._
import javax.swing.event.DocumentEvent

class SettingsDialog(applicationSettings: SnykApplicationSettingsStateService,
  projectSettings: Option[SnykProjectSettingsStateService]) {

  private val customEndpointTextField = new JTextField() with Disposable {
    override def dispose(): Unit = {}
  }

  private val organizationTextField: JTextField = new JTextField()
  private val ignoreUnknownCACheckBox: JCheckBox = new JCheckBox()
  private val additionalParametersTextField: JTextField = new ExpandableTextField()

  private val rootPanel: JPanel = new JPanel()

  setupUI()
  setupValidation()

  if (nonNull(applicationSettings)) {
    reset()
  }

  def getRootPanel: JComponent = rootPanel

  def reset(): Unit = {
    customEndpointTextField.setText(applicationSettings.getCustomEndpointUrl)
    organizationTextField.setText(applicationSettings.getOrganization)
    ignoreUnknownCACheckBox.setSelected(applicationSettings.isIgnoreUnknownCA)

    if (projectSettings.nonEmpty) {
      additionalParametersTextField.setText(projectSettings.get.getAdditionalParameters)
    }
  }

  private def setupUI(): Unit = {
    val defaultTextFieldWidth = 500

    rootPanel.setLayout(
      new UIGridLayoutManager(3, 1, new Insets(0, 0, 0, 0), -1, -1))

    val generalSettingsPanel = new JPanel(new UIGridLayoutManager(4, 3, new Insets(0, 0, 0, 0), -1, -1))
    generalSettingsPanel.setBorder(IdeBorderFactory.createTitledBorder("General settings"))

    rootPanel.add(
      generalSettingsPanel,
      new UIGridConstraints(
        0,
        0,
        1,
        1,
        UIGridConstraints.ANCHOR_CENTER,
        UIGridConstraints.FILL_BOTH,
        UIGridConstraints.SIZEPOLICY_CAN_SHRINK | UIGridConstraints.SIZEPOLICY_CAN_GROW,
        UIGridConstraints.SIZEPOLICY_CAN_SHRINK | UIGridConstraints.SIZEPOLICY_CAN_GROW,
        null,
        new Dimension(200, 200),
        null,
        0,
        false
      )
    )

    val customEndpointLabel = new JLabel("Custom endpoint:")

    generalSettingsPanel.add(
      customEndpointLabel,
      new UIGridConstraints(
        0,
        0,
        1,
        1,
        UIGridConstraints.ANCHOR_WEST,
        UIGridConstraints.FILL_NONE,
        UIGridConstraints.SIZEPOLICY_FIXED,
        UIGridConstraints.SIZEPOLICY_FIXED,
        null,
        new Dimension(120, 16),
        null,
        0,
        false
      )
    )

    generalSettingsPanel.add(
      customEndpointTextField,
      new UIGridConstraints(
        0,
        1,
        1,
        1,
        UIGridConstraints.ANCHOR_WEST,
        UIGridConstraints.SIZEPOLICY_FIXED,
        UIGridConstraints.SIZEPOLICY_FIXED,
        UIGridConstraints.SIZEPOLICY_FIXED,
        null,
        new Dimension(defaultTextFieldWidth, -1),
        null,
        0,
        false
      )
    )

    ignoreUnknownCACheckBox.setText("Ignore unknown CA")
    generalSettingsPanel.add(
      ignoreUnknownCACheckBox,
      new UIGridConstraints(
        1,
        1,
        1,
        1,
        UIGridConstraints.ANCHOR_WEST,
        UIGridConstraints.FILL_NONE,
        UIGridConstraints.SIZEPOLICY_CAN_SHRINK | UIGridConstraints.SIZEPOLICY_CAN_GROW,
        UIGridConstraints.SIZEPOLICY_FIXED,
        null,
        null,
        null,
        0,
        false
      )
    )

    val organizationLabel = new JLabel("Organization:")
    generalSettingsPanel.add(
      organizationLabel,
      new UIGridConstraints(
        2,
        0,
        1,
        1,
        UIGridConstraints.ANCHOR_WEST,
        UIGridConstraints.FILL_NONE,
        UIGridConstraints.SIZEPOLICY_FIXED,
        UIGridConstraints.SIZEPOLICY_FIXED,
        null,
        null,
        null,
        0,
        false
      )
    )

    generalSettingsPanel.add(
      organizationTextField,
      new UIGridConstraints(
        2,
        1,
        1,
        1,
        UIGridConstraints.ANCHOR_WEST,
        UIGridConstraints.SIZEPOLICY_FIXED,
        UIGridConstraints.SIZEPOLICY_FIXED,
        UIGridConstraints.SIZEPOLICY_FIXED,
        null,
        new Dimension(defaultTextFieldWidth, -1),
        null,
        0,
        false
      )
    )

    val generalSettingsSpacer = new Spacer
    generalSettingsPanel.add(
      generalSettingsSpacer,
      new UIGridConstraints(
        3,
        0,
        1,
        1,
        UIGridConstraints.ANCHOR_CENTER,
        UIGridConstraints.FILL_VERTICAL,
        1,
        UIGridConstraints.SIZEPOLICY_WANT_GROW,
        null,
        null,
        null,
        0,
        false
      )
    )

    customEndpointLabel.setLabelFor(customEndpointTextField)
    organizationLabel.setLabelFor(organizationTextField)

    if (projectSettings.nonEmpty) {
      val projectSettingsPanel = new JPanel
      projectSettingsPanel.setLayout(
        new UIGridLayoutManager(2, 3, new Insets(0, 0, 0, 0), -1, -1))
      projectSettingsPanel.setBorder(IdeBorderFactory.createTitledBorder("Project settings"))

      rootPanel.add(
        projectSettingsPanel,
        new UIGridConstraints(
          1,
          0,
          1,
          1,
          UIGridConstraints.ANCHOR_CENTER,
          UIGridConstraints.FILL_BOTH,
          UIGridConstraints.SIZEPOLICY_CAN_SHRINK | UIGridConstraints.SIZEPOLICY_CAN_GROW,
          UIGridConstraints.SIZEPOLICY_CAN_SHRINK | UIGridConstraints.SIZEPOLICY_CAN_GROW,
          null,
          new Dimension(200, 200),
          null,
          0,
          false
        )
      )

      val additionalParametersLabel = new JLabel("Additional parameters:")
      projectSettingsPanel.add(
        additionalParametersLabel,
        new UIGridConstraints(
          0,
          0,
          1,
          1,
          UIGridConstraints.ANCHOR_WEST,
          UIGridConstraints.FILL_NONE,
          UIGridConstraints.SIZEPOLICY_FIXED,
          UIGridConstraints.SIZEPOLICY_FIXED,
          null,
          new Dimension(120, 16),
          null,
          0,
          false
        )
      )

      projectSettingsPanel.add(
        additionalParametersTextField,
        new UIGridConstraints(
          0,
          1,
          1,
          1,
          UIGridConstraints.ANCHOR_WEST,
          UIGridConstraints.SIZEPOLICY_FIXED,
          UIGridConstraints.SIZEPOLICY_FIXED,
          UIGridConstraints.SIZEPOLICY_FIXED,
          null,
          new Dimension(defaultTextFieldWidth, -1),
          null,
          0,
          false
        )
      )

      additionalParametersLabel.setLabelFor(additionalParametersTextField)

      val projectSettingsSpacer = new Spacer
      projectSettingsPanel.add(
        projectSettingsSpacer,
        new UIGridConstraints(
          1,
          0,
          1,
          1,
          UIGridConstraints.ANCHOR_CENTER,
          UIGridConstraints.FILL_VERTICAL,
          1,
          UIGridConstraints.SIZEPOLICY_WANT_GROW,
          null,
          null,
          null,
          0,
          false
        )
      )

      val emptyPanel = new JPanel
      emptyPanel.setLayout(
        new UIGridLayoutManager(1, 1, new Insets(0, 0, 0, 0), -1, -1))

      rootPanel.add(
        emptyPanel,
        new UIGridConstraints(
          2,
          0,
          1,
          1,
          UIGridConstraints.ANCHOR_CENTER,
          UIGridConstraints.FILL_BOTH,
          UIGridConstraints.SIZEPOLICY_CAN_SHRINK | UIGridConstraints.SIZEPOLICY_CAN_GROW,
          UIGridConstraints.SIZEPOLICY_CAN_SHRINK | UIGridConstraints.SIZEPOLICY_CAN_GROW,
          null,
          new Dimension(200, 200),
          null,
          0,
          false
        )
      )
    }
  }

  private def setupValidation(): Unit = {
    new ComponentValidator(customEndpointTextField).withValidator(() => {

      if (isUrlValid(customEndpointTextField.getText)) {
        null
      } else {
        new ValidationInfo("Invalid custom enpoint URL.", customEndpointTextField)
      }
    }).installOn(customEndpointTextField)

    customEndpointTextField.getDocument.addDocumentListener(new DocumentAdapter() {
      protected def textChanged(documentEvent: DocumentEvent): Unit =
        ComponentValidator
          .getInstance(customEndpointTextField)
          .ifPresent((componentValidator: ComponentValidator) => componentValidator.revalidate())
    })
  }

  def organization: String = organizationTextField.getText

  def customEndpoint: String = customEndpointTextField.getText

  def isIgnoreUnknownCA: Boolean = ignoreUnknownCACheckBox.isSelected

  def additionalParameters: String = additionalParametersTextField.getText
}

File: src/main/kotlin/io/snyk/plugin/ui/SettingsDialog.kt
package io.snyk.plugin.ui

import com.intellij.openapi.Disposable
import com.intellij.openapi.components.service
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.ComponentValidator
import com.intellij.openapi.ui.ValidationInfo
import com.intellij.ui.DocumentAdapter
import com.intellij.ui.IdeBorderFactory
import com.intellij.ui.components.fields.ExpandableTextField
import com.intellij.uiDesigner.core.Spacer
import io.snyk.plugin.isProjectSettingsAvailable
import io.snyk.plugin.isUrlValid
import io.snyk.plugin.services.SnykApplicationSettingsStateService
import io.snyk.plugin.services.SnykProjectSettingsStateService
import java.awt.Dimension
import java.awt.Insets
import java.util.Objects.nonNull
import java.util.function.Supplier
import javax.swing.*
import javax.swing.event.DocumentEvent
import com.intellij.uiDesigner.core.GridConstraints as UIGridConstraints
import com.intellij.uiDesigner.core.GridLayoutManager as UIGridLayoutManager

class SettingsDialog(private val project: Project, private val applicationSettings: SnykApplicationSettingsStateService) {

    private val customEndpointTextField = UrlValidationTextField()
    private val organizationTextField: JTextField = JTextField()
    private val ignoreUnknownCACheckBox: JCheckBox = JCheckBox()
    private val additionalParametersTextField: JTextField = ExpandableTextField()

    private val rootPanel: JPanel = JPanel()

    init {
        setupUI()

        if (nonNull(applicationSettings)) {
            reset()
        }
    }

    fun getRootPanel(): JComponent = rootPanel

    fun reset() {
        customEndpointTextField.text = applicationSettings.getCustomEndpointUrl()
        organizationTextField.text = applicationSettings.getOrganization()
        ignoreUnknownCACheckBox.isSelected = applicationSettings.isIgnoreUnknownCA()

        if (isProjectSettingsAvailable(project)) {
            additionalParametersTextField.text =
                project.service<SnykProjectSettingsStateService>().getAdditionalParameters()
        }
    }

    private fun setupUI() {
        val defaultTextFieldWidth = 500

        rootPanel.layout = UIGridLayoutManager(3, 1, Insets(0, 0, 0, 0), -1, -1)

        val generalSettingsPanel = JPanel(UIGridLayoutManager(4, 3, Insets(0, 0, 0, 0), -1, -1))
        generalSettingsPanel.border = IdeBorderFactory.createTitledBorder("General settings")

        rootPanel.add(
            generalSettingsPanel,
            UIGridConstraints(
                0,
            0,
            1,
            1,
            UIGridConstraints.ANCHOR_CENTER,
            UIGridConstraints.FILL_BOTH,
            UIGridConstraints.SIZEPOLICY_CAN_SHRINK or UIGridConstraints.SIZEPOLICY_CAN_GROW,
        UIGridConstraints.SIZEPOLICY_CAN_SHRINK or UIGridConstraints.SIZEPOLICY_CAN_GROW,
        null,
        Dimension(200, 200),
        null,
        0,
        false
        ))

        val customEndpointLabel = JLabel("Custom endpoint:")
        generalSettingsPanel.add(
            customEndpointLabel,
            UIGridConstraints(
                0,
            0,
            1,
            1,
            UIGridConstraints.ANCHOR_WEST,
            UIGridConstraints.FILL_NONE,
            UIGridConstraints.SIZEPOLICY_FIXED,
            UIGridConstraints.SIZEPOLICY_FIXED,
            null,
            Dimension(120, 16),
        null,
        0,
        false
        ))

        generalSettingsPanel.add(
            customEndpointTextField,
            UIGridConstraints(
                0,
            1,
            1,
            1,
            UIGridConstraints.ANCHOR_WEST,
            UIGridConstraints.SIZEPOLICY_FIXED,
            UIGridConstraints.SIZEPOLICY_FIXED,
            UIGridConstraints.SIZEPOLICY_FIXED,
            null,
            Dimension(defaultTextFieldWidth, -1),
        null,
        0,
        false
        ))

        ignoreUnknownCACheckBox.text = "Ignore unknown CA"
        generalSettingsPanel.add(
            ignoreUnknownCACheckBox,
            UIGridConstraints(
                1,
            1,
            1,
            1,
            UIGridConstraints.ANCHOR_WEST,
            UIGridConstraints.FILL_NONE,
            UIGridConstraints.SIZEPOLICY_CAN_SHRINK or UIGridConstraints.SIZEPOLICY_CAN_GROW,
            UIGridConstraints.SIZEPOLICY_FIXED,
        null,
        null,
        null,
        0,
        false
        ))

        val organizationLabel = JLabel("Organization:")
        generalSettingsPanel.add(
            organizationLabel,
            UIGridConstraints(
                2,
            0,
            1,
            1,
            UIGridConstraints.ANCHOR_WEST,
            UIGridConstraints.FILL_NONE,
            UIGridConstraints.SIZEPOLICY_FIXED,
            UIGridConstraints.SIZEPOLICY_FIXED,
            null,
            null,
            null,
            0,
            false
        ))

        generalSettingsPanel.add(
            organizationTextField,
            UIGridConstraints(
                2,
            1,
            1,
            1,
            UIGridConstraints.ANCHOR_WEST,
            UIGridConstraints.SIZEPOLICY_FIXED,
            UIGridConstraints.SIZEPOLICY_FIXED,
            UIGridConstraints.SIZEPOLICY_FIXED,
            null,
            Dimension(defaultTextFieldWidth, -1),
        null,
        0,
        false
        ))

        val generalSettingsSpacer = Spacer()
        generalSettingsPanel.add(
            generalSettingsSpacer,
            UIGridConstraints(
                3,
                0,
                1,
                1,
                UIGridConstraints.ANCHOR_CENTER,
                UIGridConstraints.FILL_VERTICAL,
                1,
                UIGridConstraints.SIZEPOLICY_WANT_GROW,
                null,
                null,
                null,
                0,
                false
            ))

        customEndpointLabel.labelFor = customEndpointTextField
        organizationLabel.labelFor = organizationTextField

        if (isProjectSettingsAvailable(project)) {
            val projectSettingsPanel = JPanel(UIGridLayoutManager(2, 3, Insets(0, 0, 0, 0), -1, -1))
            projectSettingsPanel.border = IdeBorderFactory.createTitledBorder("Project settings")

            rootPanel.add(
                projectSettingsPanel,
                UIGridConstraints(
                    1,
                    0,
                    1,
                    1,
                    UIGridConstraints.ANCHOR_CENTER,
                    UIGridConstraints.FILL_BOTH,
                    UIGridConstraints.SIZEPOLICY_CAN_SHRINK or UIGridConstraints.SIZEPOLICY_CAN_GROW,
                    UIGridConstraints.SIZEPOLICY_CAN_SHRINK or UIGridConstraints.SIZEPOLICY_CAN_GROW,
                    null,
                    Dimension(200, 200),
                    null,
                    0,
                    false
                ))

            val additionalParametersLabel = JLabel("Additional parameters:")
            projectSettingsPanel.add(
                additionalParametersLabel,
                UIGridConstraints(
                    0,
                    0,
                    1,
                    1,
                    UIGridConstraints.ANCHOR_WEST,
                    UIGridConstraints.FILL_NONE,
                    UIGridConstraints.SIZEPOLICY_FIXED,
                    UIGridConstraints.SIZEPOLICY_FIXED,
                    null,
                    Dimension(120, 16),
                    null,
                    0,
                    false
                ))

            projectSettingsPanel.add(
                additionalParametersTextField,
                UIGridConstraints(
                    0,
                1,
                1,
                1,
                UIGridConstraints.ANCHOR_WEST,
                UIGridConstraints.SIZEPOLICY_FIXED,
                UIGridConstraints.SIZEPOLICY_FIXED,
                UIGridConstraints.SIZEPOLICY_FIXED,
                null,
                Dimension(defaultTextFieldWidth, -1),
            null,
            0,
            false
            ))

            additionalParametersLabel.labelFor = additionalParametersTextField

            val projectSettingsSpacer = Spacer()
            projectSettingsPanel.add(
                projectSettingsSpacer,
                UIGridConstraints(
                    1,
                    0,
                    1,
                    1,
                    UIGridConstraints.ANCHOR_CENTER,
                    UIGridConstraints.FILL_VERTICAL,
                    1,
                    UIGridConstraints.SIZEPOLICY_WANT_GROW,
                    null,
                    null,
                    null,
                    0,
                    false
                ))

            val emptyPanel = JPanel(UIGridLayoutManager(1, 1, Insets(0, 0, 0, 0), -1, -1))

            rootPanel.add(
                emptyPanel,
                UIGridConstraints(
                    2,
                    0,
                    1,
                    1,
                    UIGridConstraints.ANCHOR_CENTER,
                    UIGridConstraints.FILL_BOTH,
                    UIGridConstraints.SIZEPOLICY_CAN_SHRINK or UIGridConstraints.SIZEPOLICY_CAN_GROW,
                    UIGridConstraints.SIZEPOLICY_CAN_SHRINK or UIGridConstraints.SIZEPOLICY_CAN_GROW,
                    null,
                    Dimension(200, 200),
                    null,
                    0,
                    false
                ))
        }
    }

    fun getOrganization(): String = organizationTextField.text

    fun getCustomEndpoint(): String = customEndpointTextField.text

    fun isIgnoreUnknownCA(): Boolean = ignoreUnknownCACheckBox.isSelected

    fun getAdditionalParameters(): String = additionalParametersTextField.text
}

class UrlValidationTextField : JTextField(""), Disposable {

    init {
        ComponentValidator(this).withValidator(Supplier {
            val validationInfo: ValidationInfo = if (!isUrlValid(text)) {
                ValidationInfo("Invalid custom enpoint URL.", this@UrlValidationTextField)
            } else {
                ValidationInfo("")
            }

            validationInfo
        })

        this.document.addDocumentListener(object : DocumentAdapter() {
            override fun textChanged(e: DocumentEvent) {
                ComponentValidator.getInstance(this@UrlValidationTextField).ifPresent {
                    it.revalidate()
                }
            }
        })
    }

    override fun dispose() {
    }
}
